lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -1223,3 +1347,226 @@ void ConstExprEvaluator::computeConstantValues(
     numInstEvaluated = 0;
   }
 }
+
+/// Analyze the array users of an _allocateUninitialized call, to see if they
+/// are all simple things we can remove.  If so, add them all to
+/// `removableArrayInsts` and return false.  If not, return true.
+static bool getRemovableArrayInitUses(SILValue v,
+                                      SmallPtrSet<SILInstruction *, 8>
+                                          *removableArrayInsts) {
+  for (auto *use : v->getUses()) {
+    auto *user = use->getUser();
+
+    // We can always remove retain/release instructions and debug_value.
+    if (isa<StrongRetainInst>(user) || isa<StrongReleaseInst>(user) ||
+        isa<RetainValueInst>(user) || isa<ReleaseValueInst>(user) ||
+        isa<DebugValueInst>(user)) {
+      if (removableArrayInsts)
+        removableArrayInsts->insert(user);
+      continue;
+    }
+
+    // We can look through unpacking and repacking of structs that eventually
+    // turn into a retain or release.
+    if (isa<StructExtractInst>(user) || isa<StructInst>(user)) {
+      if (removableArrayInsts)
+        removableArrayInsts->insert(user);
+      if (getRemovableArrayInitUses(user->getResults()[0], removableArrayInsts))
+        return true;
+      continue;
+    }
+
+    // Oops we found an unknown use!
+    return true;
+  }
+  return false;
+}
+
+/// Try to decode the specified apply of the _allocateUninitializedArray
+/// function in the standard library.  This attempts to figure out how the
+/// resulting elements will be initialized.  This fills in the result with a
+/// lists of insts used to pass element addresses for initialization, and
+/// returns false on success.
+///
+/// Specifically, `elementsAtInit[i]` is an operand, where the associated
+/// instruction returns the address for array element `i`, and the user of that
+/// operand is a writer for that array element. For example, the user can be a
+/// store inst into that array element.
+///
+/// If `removableArrayInsts` is non-null, if decoding succeeds, and if all uses
```

Comment:
These `removeableArrayInsts` are a thing that the tensorflow branch uses to remove some array initializations in some cases. The constant evaluator itself never uses them.

I haven't thought very hard about how to deal with these in master. Leave them in, untested? Remove them from master and just maintain them in the tensorflow branch? Contrive some test for them that works in master? Anyone have thoughts?

Reply:
If we don't add support for _allocateUninitializedArray, this wouldn't be a problem right?